### DISCUSSION TITLE
Upload-only: remote not deleted

### DISCUSSION THREAD

**Discussion by Fire69** [abraunegg/onedrive#1496] [Q&A] [answered]
Hey,

This is a follow-up of: https://github.com/abraunegg/onedrive/discussions/1448

I previously was focused only on my Pictures folder and it seemed like my problems were solved, but today I looked at my Documents folder and it's not in sync.
When I add files locally, they are copied to Onedrive, but when I remove something locally, it's not deleted from Onedrive.
There are folders on Onedrive that were deleted quite some time ago.

Some info:
Onedrive runs as a service (on Ubuntu Server 20.04.2):
ExecStart=/usr/local/bin/onedrive --monitor --confdir="/home/fire69/.config/onedrive"

My config:
```
fire69@skinner:~$ onedrive --display-config
Configuration file successfully loaded
config file has been updated, checking if --resync needed
onedrive version                       = v2.4.11-6-gc9ca004
Config path                            = /home/fire69/.config/onedrive
Config file found in config path       = true
Config option 'check_nosync'           = false
Config option 'sync_dir'               = /mnt/backupdisk/backup
Config option 'skip_dir'               =
Config option 'skip_file'              = ~*|.~*|*.tmp
Config option 'skip_dotfiles'          = false
Config option 'skip_symlinks'          = false
Config option 'monitor_interval'       = 86400
Config option 'min_notify_changes'     = 5
Config option 'log_dir'                = /var/log/onedrive/
Config option 'classify_as_big_delete' = 1000
Config option 'upload_only'            = true
Config option 'no_remote_delete'       = false
Config option 'remove_source_files'    = false
Config option 'sync_root_files'        = false
Selective sync 'sync_list' configured  = true
sync_list contents:
Documents
Pictures
Videos/DV/
Business Shared Folders configured     = false
```
Part of my logfile (showing everything seems to be synced):
```
2021-May-22 20:56:58.7397447	Starting a sync with OneDrive
2021-May-22 20:56:58.7402479	Uploading differences of /mnt/backupdisk/backup
2021-May-22 21:14:02.3267184	Uploading new items of /mnt/backupdisk/backup
2021-May-22 21:14:33.7280745	Sync with OneDrive is complete
2021-May-22 21:19:33.9979369	Uploading differences of /mnt/backupdisk/backup
2021-May-22 21:35:43.3144315	Uploading new items of /mnt/backupdisk/backup
2021-May-22 21:41:12.6787781	Uploading differences of /mnt/backupdisk/backup
2021-May-22 21:58:47.4661362	Uploading new items of /mnt/backupdisk/backup
2021-May-22 22:04:16.9946428	Uploading differences of /mnt/backupdisk/backup
2021-May-22 22:21:32.0851921	Uploading new items of /mnt/backupdisk/backup
2021-May-22 22:27:01.2299496	Uploading differences of /mnt/backupdisk/backup
2021-May-22 22:43:09.7289836	Uploading new items of /mnt/backupdisk/backup
2021-May-22 22:48:38.846022	Starting a sync with OneDrive
2021-May-22 22:48:38.8484178	Uploading differences of /mnt/backupdisk/backup
2021-May-22 23:05:59.1379935	Uploading new items of /mnt/backupdisk/backup
2021-May-22 23:06:27.6407162	Sync with OneDrive is complete
2021-May-22 23:11:28.0072272	Uploading differences of /mnt/backupdisk/backup
2021-May-22 23:28:33.7020394	Uploading new items of /mnt/backupdisk/backup
2021-May-22 23:34:02.3778225	Uploading differences of /mnt/backupdisk/backup
```
As you can see it also monitors almost continuously, while my monitor_interval is set to 86400 (24h).

Any ideas?

**Comment by abraunegg** [comment]
@Fire69 
It looks like you have not restarted your client / systemd service since making changes ..

**Reply by Fire69** [reply]
I restarted the service (several times) previously to tweak some stuff after upgrading to the correct version.
Just to be sure I'll reboot the server now and check in the morning if anything changed on Onedrive.

**Reply by abraunegg** [reply]
@Fire69 
Also - remote data will *only* be deleted if:
* the client knew about the data first - ie - it was in sync before hand
* the delete is done on the local filesystem

When using upload only, you are not syncing remote data changes - so the client has no idea what has changed remotely

**Reply by Fire69** [reply]
> @Fre69
> Also - remote data will _only_ be deleted if:
> 
> * the client knew about the data first - ie - it was in sync before hand

The data that's on Onedrive was mostly uploaded by the scilion client, so after upgrading to your version, I deleted the database and a full resync was done.  So it should know the data is there now, right?

> * the delete is done on the local filesystem

I don't touch the data on Onedrive, everything there is uploaded by the client.
 
> When using upload only, you are not syncing remote data changes - so the client has no idea what has changed remotely

I don't change anything remotely

**Reply by abraunegg** [reply]
@Fire69 
> The data that's on Onedrive was mostly uploaded by the scilion client, so after upgrading to your version, I deleted the database and a full resync was done. So it should know the data is there now, right?

No .. because you are only using upload only. The client, even in a resync scenario, if you are using upload only, will only know about local files, not remote files.

If you uploaded data with the skilion client, then did an upgrade, the local db will only contain data that this client uploaded. It will not contain any record of what was prior uploaded - even though it may reside locally and remote ...

**Reply by Fire69** [reply]
So even if I deleted the db (items.sqlite3 files?) after upgrading from skilion, the new db will not contain the correct sync status after a resync?

Basically, what you're saying is that I should empty my Onedrive and let it sync from zero?

**Comment by abraunegg** [comment, marked as answer]
@Fire69 
> Basically, what you're saying is that I should empty my Onedrive and let it sync from zero?

Most likely because the client does not know what is in-sync, thus does not know that if a delete needed to occur, if the item is not in the local database, the client will not send that delete to OneDrive to delete.

For your setup, this is what I suggest:

1. As you do not change anything remotely, not bother with using `--upload-only`
2. Use `--local-first` and let the application receive the data from OneDrive which will let the application know what is fully in-sync

If you are using `--upload-only` because your online OneDrive contains data you do *not* wish to download, then I suggest the following:

1. Limit the scope of the client by using 'sync_list', 'skip_dir', 'skip_file'
2. Limit the scope of the client by using & configuring `--single-directory` to only sync, look at this directory location locally & remotely to sync data

Additionally:

* Always test with `--dry-run`
* Make sure you are running the right application version ....

**Reply by Fire69** [reply]
OK, I did what you suggested, remove the --upload-only and add the --local-first
But to make sure I'm completely in sync now I just bit the bullet and emptied my Onedrive.

Thanks again for your time and support!

**Comment by callegar** [comment]
I think that it would be quite useful to have a way to make the OneDrive identical to local (e.g. when using `--single-directory`) regardless of the sync state.
If you loose the state (e.g., because you have used `--resync` by mistake) it is way too expensive to delete everything on OneDrive and then to re-upload it, maybe just to have a bunch of files deleted on the OneDrive.

**Reply by abraunegg** [reply]
@callegar 
> I think that it would be quite useful to have a way to make the OneDrive identical to local (e.g. when using `--single-directory`) regardless of the sync state. If you loose the state (e.g., because you have used `--resync` by mistake) it is way too expensive to delete everything on OneDrive and then to re-upload it, maybe just to have a bunch of files deleted on the OneDrive.

This capability exists already, and is the default method of operation - that is, to 'sync' the state, delete files that need to be deleted, add files that need to be added. Please review the usage documentation: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md

Additionally, please review these specific options:
* `--local-first` = use local files as 'source of truth'
* `--upload-only` = upload files and delete anything that does not exist locally. If you want to *keep* files online, add `--no-remote-delete`
* `--download-only` = download all files remote and *keep* all files locally. If you want to *delete* local files that do not exist online add `--cleanup-local-files`

**Comment by callegar** [comment]
The delete of the files in OneDrive does not appear to happen at all if you have lost the local state. I have tried `--upload-only` and `--local-first` and still I could not get my OneDrive be a copy of local, because some stale files (deleted locally) were retained. I had already opened a bug about this, that was invalid as I was using an old version of `onedrive` but now I am up to date.

Seems to be the same that the user here experimented, as he says:
> I just bit the bullet and emptied my Onedrive

And
> Basically, what you're saying is that I should empty my Onedrive and let it sync from zero?
> Most likely because the client does not know what is in-sync, thus does not know that if a delete needed to occur

What I am suggesting is a way to make the remote identical to local regardless of the fact that the client has (for any reason) lost memory of what is in sync.

**Reply by abraunegg** [reply]
> What I am suggesting is a way to make the remote identical to local regardless of the fact that the client has (for any reason) lost memory of what is in sync.

Please submit a feature request

**Comment by mfergusson-ibm** [comment]
Hi was this ever implemented? I have same issue and unable to force sync even with --local-first

**Reply by abraunegg** [reply]
No feature request was ever raised

**Comment by abraunegg** [comment]
To those looking for this capability, please review the details in https://github.com/abraunegg/onedrive/issues/3649

This Feature Request implements a new config option: `--mirror-local-state`

To use this option you **must** use `--local-first`

You **cannot** use `--upload-only` as this option / mode explicitly does *not* query the online state, which is needed to be done to then purge extra files online.

#### Option Use Example:
```
onedrive --sync --verbose --local-first --mirror-local-state
```
